### PR TITLE
Quick fix for local variable shadowing

### LINKED ISSUE #34418: Compiling http_listener_method_service_test.bal in HTTP module give BIR error
Refer to this file : https://github.com/ballerina-platform/module-ballerina-http/blob/master/ballerina-tests/tests/http_listener_method_service_test.bal

As you can see, The outer service object creates new service objects here. Outer and inner service objects all have `http:Caller `called `caller` in their resource methods. Somehow, Compiler thinks that there is one `caller` variable and consider the other variable references as closures in other places.  Renaming the below `caller` references to something else will solve this issue, but the issue needs to be fixed properly.
image

Below is the error without the workaround.
```
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:164)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:445)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:126)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:126)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:403)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$2(BIROptimizer.java:108)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:108)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRTypeDefinition.accept(BIRNode.java:516)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$0(BIROptimizer.java:102)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:102)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:82)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.lambda$genBIR$0(BIRGen.java:286)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:277)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:453)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:390)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.TestCommand.execute(TestCommand.java:265)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose

Fixes #34418

This PR will check and error out redeclared symbols in function a parameters.

#### Expression bodied functions

```Ballerina
function nameArrow(int y) returns int {
    function (int, int) returns int lambda = (x, y) => x + y; // ERROR redeclared symbol 'y'
    return lambda(12, 32);
}
```

#### Lambda functions

```Ballerina
function nameArrow(int y, int z) returns int {
    function (int, int, int) returns int lambda = (x, y, z) => x + y + z; // ERROR redeclared symbol 'y'
                                                                      // ERROR redeclared symbol 'z'
    int x = 34; // this is not in a overlapping scope
    return lambda(12, 32, 33);
}
```

#### Object constructor

```Ballerina
function createService(int age) returns service object {
    public function getSum(int age) returns int;
} {

    var httpService =
    isolated service object {
        public function getSum(int age) returns int { // ERROR redeclared symbol 'age'
            return 1 + age;
        }
    };

    return httpService;
}
```

One thing to notice is since we set captured closure environment for expression bodied functions (arrow functions) in desugar. There environment scope may have things which gets after the scope. This may cause it to overlap with  scopes below. 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/01_main_function.bal [hunk 1]
```diff
@@ -44,11 +44,11 @@ public function main() returns error? {
     service object {} testServiceInMain = service object {
         resource function post resourceOne(testobserve:Caller caller, string body) {
             int numberCount = checkpanic 'int:fromString(body);
-            var sum = 0;
+            var sum1 = 0;
```
Comment (KRVPerera) [inline tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/01_main_function.bal:47]: Need to add a unit test cases like these.

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java [hunk 2]
```diff
@@ -53,8 +53,9 @@ public void testMainMethod() throws Exception {
         final String entryPointFunctionModule = DEFAULT_MODULE_ID;
         final String entryPointFunctionName = "main";
         final List<BMockSpan.CheckPoint> expectedCheckpoints = Arrays.asList(
+                new BMockSpan.CheckPoint(entryPointFunctionModule, FILE_NAME + ":20:5"),
                 new BMockSpan.CheckPoint(entryPointFunctionModule, FILE_NAME + ":22:13"),
-                new BMockSpan.CheckPoint(entryPointFunctionModule, FILE_NAME + ":24:15"),
+                new BMockSpan.CheckPoint(entryPointFunctionModule, FILE_NAME + ":25:23"),
```
Comment (KRVPerera) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java:58]: This is reverting changes done by 
https://github.com/ballerina-platform/ballerina-lang/pull/34374/commits/29899428bcc91807cbd27296c6f6554436f6a5e6

**Comment by gimantha** [comment]
@KRVPerera Shall we send the PR to `stage-swan-lake-rc1` ?

**Comment by chiranSachintha** [comment]
For this kind of example, it doesn't give a compile-time error.
Isn't a problem
```ballerina
public function main() {
    int a = 10;
    var x = function (function(int a) l) {};
}
```

**Comment by KRVPerera** [comment]
> ```ballerina
> public function main() {
>     int a = 10;
>     var x = function (function(int a) l) {};
> }
> ```

It does not shadow a variable right this is just the shape. 
```Ballerina
public function main() {
    int a = 10;
    var x = function (function(int) l) {};
}
```

**Comment by chiranSachintha** [comment]
```ballerina
//Example 1
public function main() {
    int a = 10;
    var x = function (function(int a = 10) l) {};
}

//Example 2
public function main() {
    int a = 10;
    var x = function (function(int a, int b = a) l) {};
}
```
But for these kind of instances, we cannot ignore the name of the parameter. (So we need to define the parameters within their scope. then according to the spec, `it is a compile error if an identifier is declared with block-scope and its scope overlaps with the scope of another declaration of the same identifier in the same symbol space also with block-scope. ` It is a compile-time error)

**Comment by KRVPerera** [comment]
Referring the issue https://github.com/ballerina-platform/ballerina-spec/issues/873 here

**Comment by KRVPerera** [comment]
> 

I fixed this
```LogTalk
ERROR [t1.bal:(3:32,3:42)] redeclared symbol 'a'
ERROR [t1.bal:(9:32,9:37)] redeclared symbol 'a'
```

**Comment by KRVPerera** [comment]
> ```ballerina
> //Example 1
> public function main() {
>     int a = 10;
>     var x = function (function(int a = 10) l) {};
> }
> 
> //Example 2
> public function main() {
>     int a = 10;
>     var x = function (function(int a, int b = a) l) {};
> }
> ```
> 
> But for these kind of instances, we cannot ignore the name of the parameter. (So we need to define the parameters within their scope. then according to the spec, `it is a compile error if an identifier is declared with block-scope and its scope overlaps with the scope of another declaration of the same identifier in the same symbol space also with block-scope. ` It is a compile-time error)

I was able to fix this.

```LogTalk
ERROR [t1.bal:(3:32,3:42)] redeclared symbol 'a'
ERROR [t1.bal:(9:32,9:37)] redeclared symbol 'a'
```

**Review by hasithaa — APPROVED** [review]

**Comment by chiranSachintha** [comment]
Consider the following examples
```ballerina
public function main() {
    int x = 10;
    int x = "";
}
```
This gives error as `redeclared symbol 'x'`

```ballerina
public function main() {
    int x = 10;
    var y = function (string x = 13) {};
}
```
This gives error as
```
ERROR [test3.bal:(9:23,9:36)] redeclared symbol 'x'
ERROR [test3.bal:(9:34,9:36)] incompatible types: expected 'string', found 'int'

```
Isn't it a problem? If we give a second error(type incompatible error) it means that the symbols are defined within the scope correctly but the type mismatch.

**Pull request merged**